src/components/DownloadButton.tsx

```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import { set } from 'idb-keyval';
+import GetAppIcon from '@material-ui/icons/GetApp';
+import { videoOrArticle, VideoData, ArticleData } from '../store/resourcesSlice'
+import { videoStore } from '../index'
+import Button from '@material-ui/core/Button';
+import { useDispatch } from 'react-redux';
+import { setResourceIsCached } from '../store/resourcesSlice';
+
+type DownloadButtonProps = {
+  id: number,
+  videoData?: VideoData | ArticleData,
+}
+
+const DownloadButton = ({ id, videoData }: DownloadButtonProps) => {
+  const dispatch = useDispatch();
+
+  async function downloadVideo(resId, videoStore, videoData) {
+    let downloadUrl = videoOrArticle(videoData) ? videoData.downloadUrl : null
+    if (downloadUrl) {
+      let blob = await fetch(downloadUrl).then(res => res.blob())
+      set(resId, blob, videoStore).then(() => dispatch(setResourceIsCached({
```

Comment:
We should also add a `.catch` to this to handle the case where we're unable to cache the video data. That way we can display a message to the user that it failed, since this is usually a sign that either:
- They're out of memory and can no longer download more videos (need to delete some videos / clear cache)
- They're in private browsing mode and might not be able to save anything to the cache at all

If it's not too hard, it would be great to figure out the exact error reason and display that to the user!

Reply:
Ooh I see! I'll add the catch case -- I'll prob have to double check how idb-keyval returns errors!